### ISSUE TITLE
Specifying AnalogLinear or AnalogConv2d to run on cuda

### ISSUE THREAD

**Issue by chaeunl** [IBM/aihwkit#70]
Thank you for you guys to give detailed replies for any issues.

I am now composing a neural network which consists of AnalogLinear modules. But, when I am trying to specify the device, I have to specify each AnalogLinear module to run on cuda, using _.cuda()_.  As you know, PyTorch modules are automatically run on cuda when users specify the network to run on cuda. But, with AnalogLayer or AnalogConv2d, without specifying each layer to use cuda, the error message occurs.

 I am not sure that this improvement requires efforts. It would be convenient if each module automatically run on cuda by specifying it on the network level.

**Comment by diego-plan9** [comment]
Could you provide a code snippet with the basics on how you create your network, and the error that you hint at in the issue description? In `0.2.0` we [introduced `aihwkit.nn.AnalogSequential`](https://aihwkit.readthedocs.io/en/latest/using_pytorch.html#using-analog-layers-as-part-of-other-modules) as an optional replacement of `nn.Sequential` in the hopes of alleviating a bit the need for moving the children layers individually to cuda which might help:

```python
from aihwkit.nn import AnalogLinear, AnalogSequential

model = AnalogSequential(
    AnalogLinear(10, 20)
)
model.cuda()
```

We resorted to adding the new class as the stock torch modules seem to only modify `Parameters` and `Buffers` when applying recursive functions (such as calling `.cuda()` on the parent) - and at the moment our `Tiles` are not among those. There is still room for improvement and tuning, so looking forward to see if it fits your current scenario and feedback!

**Comment by chaeunl** [comment]
Oh! I have not noticed AnalogSequential is embodied. I just composed a network using nn.Sequential module. Then, at now, tiles are run on CPU even the AnalogLayers are specified to use cuda? Also, I am wondering that why the amount of memory allocation is quite high with simple MLP (784-256-128-10); it takes about more than 5GB memory in GPU.

**Comment by diego-plan9** [comment]
Would it be possible to have a snippet showing how you are creating the network and moving it to cuda, for easier discussion and help us identify potential issues?

Tiles _should_ end up in the GPU in all cases if using `.cuda()`:
* if using `nn.Sequential`, `nn.Sequential.cuda()` would move the weights and biases tensors (but not the tiles, unfortunately) of the children analog layers to the GPU - and we have a fall back in the optimizer for also moving the tile at that point in time if it detects the analog layer has its weights and biases in the GPU but not the tile.
* if using `AnalogSequential`, the moving to the GPU of the children layers (including their tiles) should happen in the "natural" way at `AnalogSequential.cuda()`.
* if using `AnalogLinear.cuda()` directly, the moving should also happen at that time.

>  Also, I am wondering that why the amount of memory allocation is quite high with simple MLP (784-256-128-10); it takes about more than 5GB memory in GPU.

Would you have a comparison of the size of the same network without analog layers, for reference and helping us investigate the potential performance issues?

**Comment by chaeunl** [comment]
> Would it be possible to have a snippet showing how you are creating the network and moving it to cuda, for easier discussion and help us identify potential issues?
```python
    model = nn.Sequential(
            AnalogLinear(layers_size[0], layers_size[1], True,
                         rpu_config=rpu_config).cuda(),
            nn.Sigmoid(),
            AnalogLinear(layers_size[1], layers_size[2], True,
                         rpu_config=rpu_config).cuda(),
            nn.Sigmoid(),
            AnalogLinear(layers_size[2], layers_size[3], True,
                         rpu_config=rpu_config).cuda(),
        ).cuda()
```

> Would you have a comparison of the size of the same network without analog layers, for reference and helping us investigate the potential performance issues?

1) When I made the network using the above code, here is the info:
```
Epoch 1    of 1
   Step [6000  /60000 ], Loss: 2.3439  
   ...
   Step [60000 /60000 ], Loss: 2.1908  
 => Train loss/acc.: 2.1908  /17.9633  Val loss/acc.: 1.9478  /29.9800 
 => GPU Memory Usage:2575.8125 (GiB)
 => Elapsed Time (s) = 78.29633212089539
```

2) Replacing the `nn.Sequential()` with `AnalogSequential()`:
```
Epoch 1    of 1   
   Step [6000  /60000 ], Loss: 2.3491  
   ...
   Step [60000 /60000 ], Loss: 2.1832  
 => Train loss/acc.: 2.1832  /18.7400  Val loss/acc.: 1.9729  /27.9900 
 => GPU Memory Usage:2575.8125 (GiB)
 => Elapsed Time (s) = 79.26095652580261
```

I wrongly reported before the memory usage :( 
I think that these two cases make no difference in usage of memory and elapsed time.

**Comment by diego-plan9** [comment]
Thanks for the detailed info @chaeunl ! To recap:
* can we consider that the original issue has been resolved, or are you still experiencing errors? I'm inferring it is the case (basically based on the memory bump you reported for 1 & 2), but please let me know otherwise so we can take action.
  You can double-check if the tile is really using cuda via something similar to (for the first children tile in your model):
  ```
  list(model.children())[0].analog_tile.is_cuda
  ```
  It should be `True` after `.cuda()` if solving it via options 2 or 3 from https://github.com/IBM/aihwkit/issues/70#issuecomment-724672347; and after `regroup_param_groups()` if solving it via option 1.

* thank you as well for the "benchmarking" on the GPU memory usage! The new numbers look more reasonable indeed - hopefully @maljoras can verify they are within the expected range.

**Comment by maljoras** [comment]
The CUDA implementation for the analog layers  indeed needs a number of additional copies and buffers, at least 1 for each activation and at least one copy of the weight (depending on the choice of the parameter. e.g weight noise etc). However, @chaeunl, the number 2500 GB (!) seems to be too high though, are you sure you are reporting GB and not MB? 2.5GB might be more reasonable. Currently, we also have an additional copy of the weights as the parameter, which we actually do not need, but in your case weights are tiny anyway. Also, note that pytorch might not free memory immediately, but will overwrite those buffers later I suppose, so the amount of memory reported may be considerably more than the memory actual needed.

**Comment by chaeunl** [comment]
Oops! Yes, I misprinted the memory usage. 2500MiB is right (The capacity of our gpu is 24 GiB..). Indeed, we are using PyTorch 1.8.0 ver. which has not been released and optimized yet. So, it might allocate more memory. The way that I adopted to observe memory usage is I guess same as the way to call `nvidia-smi` from the console. I think if someone who use PyTorch 1.7.0 ver. might confirm that the memory is less allocated. Thank you for detailed replies. @maljoras and @diego-plan9  :)
